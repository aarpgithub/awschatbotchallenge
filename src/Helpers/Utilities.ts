import { isNil } from "lodash";
import { createHash } from "crypto";
import traverse = require("traverse");
import { union as lodash_union, flatten as lodash_flatten } from "lodash";

/**
 * Remove any keys with undefined or null values from obj and return obj.  
 * !! Mutates argument !!
 * 
 * example: 
 *  const obj = {a: null, b: 1}
 *  const obj2 = excludeNil(obj)
 *  console.log(obj2)
 * -> "{ b: 1}"
 * @param obj 
 */
export function excludeNil<T>(obj: T): T {
  for (const key in obj) {
    if (isNil(obj[key])) {
      delete obj[key];
    }
  }
  return obj;
}

export const sha256 = (str: string): string => {
  const hash = createHash("sha256");
  hash.update(str);
  return hash.digest("hex");
};

export const sha256checksum = (obj: Object): string => {
  return sha256(JSON.stringify(obj));
};

export const removeKeysRecursive = (obj: any, keys: Array<String>) => {
  traverse(obj).forEach(function(e: any) {
    if (this.key && containsKey(this.key, keys)) {
      this.remove(true);
    }
  });
};

export const containsKey = (key: string, keys: Array<String>) => {
  if (keys.indexOf(key) > -1) {
    return true;
  }
  return false;
};

export const snooze = (ms: number) => {
  return new Promise(resolve => setTimeout(resolve, ms));
};

export const union = lodash_union;
export const flatten = lodash_flatten;
