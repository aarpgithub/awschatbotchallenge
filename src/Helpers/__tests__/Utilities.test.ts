import { excludeNil, sha256checksum } from "../Utilities";

it("excludeNil should remove null/undefined values from top-level of object", () => {
  const testcase = { a: 1, b: "1", c: null, d: undefined, e: { a: 1 } };
  expect(excludeNil(testcase)).toMatchSnapshot();
});

it("sha256checksum of object should produce expected checksum value", () => {
  const testcase = { a: 1, b: "1", c: null, d: undefined, e: { a: 1 } };
  expect(sha256checksum(testcase)).toMatchSnapshot();
});
