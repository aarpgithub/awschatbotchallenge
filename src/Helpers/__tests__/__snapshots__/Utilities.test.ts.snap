// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`excludeNil should remove null/undefined values from top-level of object 1`] = `
Object {
  "a": 1,
  "b": "1",
  "e": Object {
    "a": 1,
  },
}
`;

exports[`sha256checksum of object should produce expected checksum value 1`] = `"f495c9d58385356851afed14daf8518ef6adbf8beea1b58a0014a4a06389a777"`;
